File: bishi/9-27华为/a.py
from collections import deque

def bfs(start, adj, color, groups):
    q = deque([start])
    color[start] = 0 
    groups[0].append(start) 

    while q:
        u = q.popleft()
        for v in adj[u]:
            if color[v] == -1: 
                color[v] = 1 - color[u]  
                groups[color[v]].append(v) 
                q.append(v)
            elif color[v] == color[u]: 
                return False  
    return True

def find_groups(n, adj):
    color = [-1] * n 
    res = []

    for start in range(n):
        if color[start] == -1: 
            groups = [[] for _ in range(2)] 
            if not bfs(start, adj, color, groups):
                return -1  
            res.append(groups)

    ans1, ans2 = set(), set()
    for groups in res:
        ans1.update(groups[0])
        ans2.update(groups[1])

    ans1 = sorted(ans1)
    ans2 = sorted(ans2)

    return ans1, ans2

if __name__ == "__main__":
    n = int(input())
    adj = []

    for _ in range(n):
        relations = list(map(int, input().strip().split()))
        adj.append(relations)

    result = find_groups(n, adj)
    if result == -1:
        print(-1)
    else:
        group1, group2 = result
        print(' '.join(map(str, group1)))
        print(' '.join(map(str, group2)))

File: bishi/9-27华为/a.cpp
#include <bits/stdc++.h>

const int inf = 0x3f3f3f3f;
using i64 = long long;

template<typename T>
std::vector<T> input() {
    std::vector<T> a;
    T s;
    while(std::cin >> s) {
        a.push_back(s);
        if(std::cin.get() != ' ') break;
    }
    return a;
}

bool bfs(int start, std::vector<std::vector<int>> &adj, std::vector<int> color, std::vector<std::vector<int>> groups)  {
    std::queue<int> q;
    q.push(start);
    color[start] = 0;
    groups[0].push_back(start);

    while (q.size()) {
        int u = q.front();
        q.pop();
        for (auto v : adj[u]) {
            if (color[v] == -1) {
                color[v] = 1 - color[u];
                groups[color[v]].push_back(v);
                q.push(v);
            } else if (color[v] == color[u]) {
                return false;
            }
        }
    }    
    return true;
}

void solve() {
    int n;
    std::cin >> n;
    std::vector<int> st(n, -1);
    std::vector<std::vector<int>> adj;
    for (int i = 0; i < n; ++i) {
        auto t = input<int>();
        adj.push_back(t);
    }

    std::vector<std::vector<std::vector<int>>> res;
    for (int i = 0; i < n; ++i) {
        if (st[i] == -1) {
            std::vector<std::vector<int>> a(2);
            if (!bfs(i, adj, st, a)) {
                std::cout << "-1\n";
                return;
            }
            res.push_back(a);
        }
    }
    std::set<int> ans1, ans2;
    for (auto g : res) {
        for (auto x : g[0]) {
            ans1.insert(x);
        }

        for (auto x : g[1]) {
            ans2.insert(x);
        }
    }
    for (auto x : ans1) {
        std::cout << x << " ";
    }
    std::cout << "\n";
    for (auto x : ans2) {
        std::cout << x << " ";
    }
}

int main() {
    std::cin.sync_with_stdio(false);
    std::cin.tie(0);
    
    solve();

    return 0;
}
